toucan_connectors/s3/s3_connector.py

```diff
@@ -0,0 +1,164 @@
+
+        forged_url = f's3://{access_key}:{access_secret}@{absolute_path}'
+
+        return forged_url
+
+    def _retrieve_data(
+        self,
+        data_source: S3DataSource,
+        offset: int = 0,
+        limit: int | None = None,
+    ) -> pd.DataFrame | TextFileReader:
+        credentials = self._get_assumed_sts_role()['Credentials']
+
+        # See the documentation here for the schema :
+        # https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sts/client/assume_role.html
+        access_key = credentials['AccessKeyId']
+        access_secret = credentials['SecretAccessKey']
+        session_token = credentials['SessionToken']
+
+        reader_kwargs = {
+            **data_source.reader_kwargs,
+            'preview_nrows': limit,
+            'preview_offset': offset,
+        }
+        fetcher_kwargs = data_source.fetcher_kwargs
+        client_kwargs = fetcher_kwargs.get('client_kwargs', {})
+        fetcher_kwargs = {
+            **fetcher_kwargs,
+            'client_kwargs': {**client_kwargs, 'session_token': session_token},
+        }
+
+        session = boto3.Session(
+            aws_access_key_id=access_key,
+            aws_secret_access_key=access_secret,
+            aws_session_token=session_token,
+        )
+        s3 = session.client('s3')
+        paginator = s3.get_paginator('list_objects_v2')
+        pages = paginator.paginate(Bucket=self.bucket_name, Prefix=self.prefix)
+        pattern = re.compile(rf'{data_source.file}')
+
+        dfs = []
+        # We loop over pages to fetch for available file's paths as object keys
+        for page in pages:
+            for obj_sum in page.get('Contents', []):
+                file_path = obj_sum['Key']
+                # We check the regex match pattern for the given file path
+                if pattern.match(file_path):
+                    s3_uri = self._forge_url(
+                        access_key=access_key,
+                        access_secret=access_secret,
+                        session_token=session_token,
+                        file=file_path,
+                    ).rstrip('$')
+                    df = DataSource(
+                        uri=s3_uri,
+                        reader_kwargs=reader_kwargs,
+                        fetcher_kwargs=fetcher_kwargs,
+                    ).get_df()
+                    df['__filename__'] = file_path
+                    dfs.append(df)
+
+        return pd.concat(dfs)
+
+    @cached_property
+    def _sts_assumed_role(self) -> dict:
+        sts_client = boto3.client(
+            'sts',
+            aws_access_key_id=self.sts_access_key_id,
+            aws_secret_access_key=self.sts_secret_access_key,
+        )
+        return sts_client.assume_role(
+            RoleArn=self.role_arn,
+            RoleSessionName='toucantoco',
+            ExternalId=self.workspace_id,
+        )
+
+    def _get_assumed_sts_role(self) -> dict:
+        with suppress(Exception):
```

Comment:
```suggestion
        with suppress(KeyError):
```
to me more explicit ?

Reply:
we don't want to be specific here, whatever exception is raised we want to suppress it
i don't actually think anything else could be raised for now but it seems more explicit to me this way, wyt?

Reply:
let's keep it simple then with only "Exception" 👌🏽